### PR TITLE
CXX-861 Add example iterating cursor and getting data out

### PR DESCRIPTION
**Pull request by xdg**
I wrote out the types long-hand instead of using 'auto' into order
to better educate readers about what types are being used.

I also omitted most error handling to keep the focus on working with
the cursor, documents, arrays and elements.

### CODE REVIEW

**Comment by xdg** [comment]
PTAL: @jrassi, @acmorrow

**Comment by xdg** [comment]
Updated to fix a failing type conversion on old Clang.

@jrassi please look this over when you get a chance.  I'd like to merge it and close the ticket.
Thanks!

File: examples/mongocxx/get_values_from_documents.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <mongocxx/instance.hpp>
+#include <mongocxx/client.hpp>
+
+using namespace bsoncxx;
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest pulling in the types you need from bsoncxx individually, instead of adding a using-directive.  This suggestion is mostly for consistency with the other examples in this directory.  Note also that the Google C++ style guide bans using-directives entirely (see <https://google.github.io/styleguide/cppguide.html#Namespaces), although it looks like the driver uses them moderately.  Mostly in test code, it seems.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+//       {
+//         "uid" : 413098706,
+//         "status" : 2,
+//         "msg": "..."
+//       },
+//       ...
+//     ]
+// }
+
+// Construct a document in the format of 'messagelist'
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest ending comments in periods.  Ditto throughout.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+document::value new_message(int64_t uid, int32_t status, std::string msg) {
+    builder::stream::document doc;
+    return doc << "uid" << uid << "status" << status << "msg" << msg << builder::stream::finalize;
+}
+
+// Insert a document into the database
+void insert_test_data(mongocxx::collection& coll) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest changing `coll` to be a pointer type, instead of a reference type.  This is just straight from the Google C++ style guide; see <https://google.github.io/styleguide/cppguide.html#Reference_Arguments>.

Same thing for the declaration of `iterate_documents()`.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    builder::stream::document builder{};
+    builder << "messagelist" << builder::stream::open_array
+            << new_message(413098706, 3, "Lorem ipsum...")
+            << new_message(413098707, 2, "Lorem ipsum...")
+            << new_message(413098708, 1, "Lorem ipsum...") << builder::stream::close_array;
+
+    document::value doc = builder << builder::stream::finalize;
+
+    // Normally, one should check the return value for success
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: I'm of the opinion that all examples should feature production-quality error handling, though I see your dissent in #519.  Perhaps let's discuss further offline.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    coll.insert_one(doc.view());
+}
+
+// Iterate over contents of messagelist
+void iterate_messagelist(document::element& ele) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest declaring `ele` as const.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    // Check validity && type before trying to iterate
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Nit: suggest spelling out "and" instead of using "&&" in a comment.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    if (ele.type() == type::k_array) {
+        array::view subarray{ele.get_array().value};
+        for (array::element msg : subarray) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: I'm somewhat conflicted as far as whether a detailed exploration of the document_view/element methods belongs here, or in an existing bsoncxx example like getting_values.cpp.  What do you think?

File: examples/mongocxx/get_values_from_documents.cpp [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            // Check correct type before trying to access elements.
+            // Only print out fields if they exist; don't report missing fields.
+            if (msg.type() == type::k_document) {
+                document::view subdoc{msg.get_document().value};
+                if (subdoc["uid"] && subdoc["uid"].type() == type::k_int64) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest pulling out the `subdoc["uid"]` element into a local variable (perhaps `uidEle`?), instead of calculating it a few times.  Same with the other elements in this block.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            } else {
+                std::cout << "Message is not a document" << std::endl;
+            }
+        }
+    } else {
+        std::cout << "messagelist is not an array" << std::endl;
+    }
+}
+
+// Print document parts to STDOUT
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Nit: suggest "standard out" instead.  STDOUT makes me think of Perl.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+void print_document(const bsoncxx::document::view& doc) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Nix the `bsoncxx::` prefix here, as you have a using-directive in scope.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    // Define an element view for inspecting different fields
+    document::element ele{};
+
+    // Extract _id element as a string
+    ele = doc["_id"];
+    if (ele.type() == type::k_oid) {
+        std::string oid = ele.get_oid().value.to_string();
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Suggest omitting the to_string() call, for consistency with how you print `msg` at 74, and because you don't need an owned string here.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        std::cout << "OID: " << oid << std::endl;
+    } else {
+        std::cout << "Error: _id was not an object ID." << std::endl;
+    }
+
+    // Extract "messagelist" element, which could be the 'invalid' (false)
+    // element if it doesn't exist
+    ele = doc["messagelist"];
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: I find it slightly more stylish / easier to read to declare a new document::element variable here with a descriptive variable name, instead of reusing the one declared above at 88.  Suggest something like: deleting 88, replacing 91 with `document::element idEle = doc["_id"];`, and replacing 101 with `document::element messageListEle = doc["messageList"];`.

**Comment by jrassi** [comment]
Left a round of feedback, which is mostly style nits.  See also feedback at [get_values_from_documents.cpp:62](https://github.com/mongodb/mongo-cxx-driver/pull/518#discussion_r77737522), where I express uncertainty about where this example should live.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 3 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: It could be `const` -- we don't modify the *object*, though we do modify data on MongoDB.  My view on reference/pointer is "pass reference when it can't be null" (I think that's one of Meyer's prescriptions.)

Comment (acmorrow) [inline examples/mongocxx/get_values_from_documents.cpp]: @jrassi @xdg - There is a general kernel team avoidance of using non-const reference types for parameters, since it makes it hard at the call site to know whether a given parameter may be being changed by the call. If you make "in parameters" by-pointer, then it is clearer at the call sites which parameters are expected to be changed. I'm not taking a position either way on whether we want this guidance to apply in the C++11 driver, at least not quite yet.

Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: I agree it's a good convention. I suspect when I wrote this, I hadn't internalized the importance of "const" yet and was just getting the code to work.  I'll go through and amend this and other parameters accordingly.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 1 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 2 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 3 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: I see what happened.  Things like "insert_one" aren't currently const methods.  I think that should change as the *objects* are const.  Yes, they carry out operations with side effects that mutate database state, but that's entirely external to the C++ code.

Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: I'm going to leave these as non-const references as making them pointers implies mutation, which we don't do either.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 4 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: I've put it on the team meeting discussion list

File: examples/mongocxx/get_values_from_documents.cpp (hunk 5 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 6 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 7 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: The original user question has the data model described so the example code answered their question which included getting elements out of the nested array.  I think giving end-to-end examples like this are useful and will forestall additional questions.  Saying it differently: I think task-focused examples may be more immediately useful to users than narrow API examples that require them to stitch together concepts from multiple example files.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 8 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Good point given that `operator[]` is O(N)!  Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 9 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 10 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done

File: examples/mongocxx/get_values_from_documents.cpp (hunk 11 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: I think it can't be done, as `bsoncxx::oid` doesn't provide a helper for streaming to output and doesn't provide a string_view option.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 12 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done

**Comment by xdg** [comment]
@jrassi @acmorrow I've pushed a commit with changes. PTAL.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 3 above)
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: I'm not in agreement about making insert_one const, but I'd like to defer discussion of that.

I do feel more strongly about having a consistent rule across the codebase for reference versus pointer function parameters, though.  From a quick-and-dirty grep of the source I can find 1 file containing function declarations/definitions with non-const reference parameters, versus 38 files containing function declarations/definitions with non-const pointer parameters, and I'd love to maintain conformance with the existing style where possible.  I'd be happy take a look at the Meyers' recommendation and consider it for the future, but I'd say the project currently inherits the "All parameters passed by reference must be labeled const" rule from claiming to follow Google C++ style.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 7 above)
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Acknowledged.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 13]
```diff
@@ -18,7 +18,22 @@
...
+using bsoncxx::builder::stream::finalize;
+using bsoncxx::builder::stream::open_array;
+using bsoncxx::type::k_array;
+using bsoncxx::type::k_document;
+using bsoncxx::type::k_int32;
+using bsoncxx::type::k_int64;
+using bsoncxx::type::k_oid;
+using bsoncxx::type::k_utf8;
+
+using doc_builder = bsoncxx::builder::stream::document;
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Instead of seeing type aliases used for these, I'd much rather either see any of the following, in decreasing order of preference: 1) spelling them out, like the other mongocxx examples do, 2) putting back the mongocxx using-declaration that you had in your first patch set, or 3) using auto.  We should change the names for these classes if we dislike them, but until then I don't like encouraging users to use different names for them.

Not sure if I was unclear in my earlier suggestion, apologies if so.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 14]
```diff
@@ -34,44 +49,47 @@ using namespace bsoncxx;
...
-    // Check validity && type before trying to iterate
-    if (ele.type() == type::k_array) {
-        array::view subarray{ele.get_array().value};
-        for (array::element msg : subarray) {
+// Iterate over contents of messagelist.
+void iterate_messagelist(const doc_element& ele) {
+    // Check validity and type before trying to iterate.
+    if (ele.type() == k_array) {
+        array_view subarray{ele.get_array().value};
+        for (array_element& msg : subarray) {
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Nit: suggest making `msg` a non-reference type (`element` is only 16 bytes and is trivially destructible) or a const reference, if you don't need to modify it.

File: examples/mongocxx/get_values_from_documents.cpp [hunk 15]
```diff
@@ -82,25 +100,22 @@ void iterate_messagelist(document::element& ele) {
...
+    doc_element id_ele = doc["_id"];
+    if (id_ele.type() == k_oid) {
+        std::string oid = id_ele.get_oid().value.to_string();
         std::cout << "OID: " << oid << std::endl;
     } else {
         std::cout << "Error: _id was not an object ID." << std::endl;
     }
 
     // Extract "messagelist" element, which could be the 'invalid' (false)
     // element if it doesn't exist
```
Comment (jrassi) [inline examples/mongocxx/get_values_from_documents.cpp]: Nit: period here, and on 125, and on 128.

**Comment by jrassi** [comment]
One more round, all style nits this time.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 13 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 14 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: Done.

File: examples/mongocxx/get_values_from_documents.cpp (hunk 15 above)
Comment (xdg) [inline examples/mongocxx/get_values_from_documents.cpp]: done.

**Comment by jrassi** [comment]
Let's discuss the thread at <
https://github.com/mongodb/mongo-cxx-driver/pull/518#discussion_r78251165>
at some point (I also think it's possible that you missed my last comment
on it, I'm personally finding GitHub PRs still a bit hard to follow).
